### PR TITLE
Require explicit return types on public methods

### PR DESCRIPTION
**Pull request by ceedubs**
@non [mentioned](https://github.com/non/cats/issues/291#issuecomment-99097551) that he would like to see this requirement. I don't know that I think we need it _everywhere_, but I do like that it catches cases where we accidentally forget to add a return type, as happened [here](https://github.com/non/cats/pull/302#discussion_r29667659) and some other places.

This pull requests adds a Scalasytle rule that enforces this constraint. It also updates the existing code to pass the Scalastyle rule.

Note: Scalastyle is producing false-positives for local defs.

### CODE REVIEW

**Comment by non** [comment]
How did you feel about this process? I think adding the types makes the code a bit more annoying to write but more readable.

I'm happy to give this :+1: but I worry about being unilateral re my stance on annotations.

**Comment by ceedubs** [comment]
@non I'd say your assessment of "adding the types makes the code a bit more annoying to write but more readable" is a fair one.

The only part that didn't seem worth it was defining the RuleSets for the laws. After updating a few, I wrote a vim macro to do the rest, otherwise it would have been really painful :). I think this particular case could be remedied with a helper method for creating RuleSets.

I'm definitely interested in hearing what others have to say.

**Comment by non** [comment]
Right, it seems like maybe a factory constructor with named arguments (and possibly defaults) would work well as a substitute for `new RuleSet`.

**Comment by ceedubs** [comment]
@non I switched most of them over to use Discipline's `DefaultRuleSet` constructor. The six tests that have multiple parents are still using `new RuleSet`. They could be switched over, but because RuleSet is a path-dependent type, I think I would have to introduce a new base class for laws or something. I thought I would hold off on that for now at least.

File: laws/src/main/scala/cats/laws/discipline/AlternativeTests.scala [hunk 1]
```diff
@@ -19,13 +20,16 @@ trait AlternativeTests[F[_]] extends ApplicativeTests[F] with MonoidKTests[F]  {
...
+      val name: String = "alternative"
+      val bases: Seq[(String, RuleSet)] = Nil
+      val parents: Seq[RuleSet] = Seq(monoidK[A], applicative[A,B,C])
+      val props: Seq[(String, Prop)] = Seq(
+        "left distributivity" -> forAll((fa1: F[A], fa2: F[A], fab: A => B) =>
+          laws.alternativeLeftDistributivity[A,B](fa1, fa2, fab)),
+        "right distributivity" -> forAll((fa: F[A], fab1: F[A => B], fab2: F[A => B]) =>
+          laws.alternativeRightDistributivity[A,B](fa, fab1, fab2)),
+        "right absorption" -> forAll((fab: F[A => B]) =>
+          laws.alternativeRightAbsorption[A,B](fab))
```
Comment (fthomas) [inline laws/src/main/scala/cats/laws/discipline/AlternativeTests.scala]: Adding explicit parameters for the laws seems unnecessary to me. What is the reason for this change?

Comment (ceedubs) [inline laws/src/main/scala/cats/laws/discipline/AlternativeTests.scala]: Yeah I agree. For some reason it stopped compiling until I did that. I didn't look into it much, and it's possible that I did something silly.

**Comment by fthomas** [comment]
In some places this seems to add unnecessary noise (like in `def laws: ApplyLaws[F] = ApplyLaws[F]`), but overall I like this. :+1:

**Comment by non** [comment]
:+1:
